codex/stores/repostore/coders.nim

```diff
@@ -0,0 +1,53 @@
+## Nim-Codex
+## Copyright (c) 2024 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+##
+
+import std/sugar
+import pkg/libp2p/cid
+import pkg/serde/json
+import pkg/stew/byteutils
+import pkg/stew/endians2
+
+import ./types
+import ../../errors
+import ../../merkletree
+import ../../utils/json
+
+proc encode*(t: QuotaUsage): seq[byte] = t.toJson().toBytes()
+proc decode*(T: type QuotaUsage, bytes: seq[byte]): ?!T = T.fromJson(bytes)
+
+proc encode*(t: BlockMetadata): seq[byte] = t.toJson().toBytes()
+proc decode*(T: type BlockMetadata, bytes: seq[byte]): ?!T = T.fromJson(bytes)
+
+proc encode*(t: LeafMetadata): seq[byte] = t.toJson().toBytes()
+proc decode*(T: type LeafMetadata, bytes: seq[byte]): ?!T = T.fromJson(bytes)
+
+proc encode*(t: DeleteResult): seq[byte] = t.toJson().toBytes()
+proc decode*(T: type DeleteResult, bytes: seq[byte]): ?!T = T.fromJson(bytes)
+
+proc encode*(t: StoreResult): seq[byte] = t.toJson().toBytes()
+proc decode*(T: type StoreResult, bytes: seq[byte]): ?!T = T.fromJson(bytes)
+
+proc encode*(i: uint64): seq[byte] =
+  @(i.toBytesBE)
+
+proc decode*(T: type uint64, bytes: seq[byte]): ?!T =
+  if bytes.len >= sizeof(uint64):
+    success(uint64.fromBytesBE(bytes))
+  else:
+    failure("Not enough bytes to decode `uint64`")
+
+proc encode*(i: int): seq[byte] = cast[uint64](i).encode
```

Comment:
Avoid using `cast` as much as possible. Nim has safe type conversion and unsafe type conversion, `cast` is explicitly unsafe.

Just doing:

```suggestion

proc encode*(i: int): seq[byte] = uint64(i).encode

```

should suffice.

Reply:
I already had that discussion with @emizzle link: https://github.com/codex-storage/nim-codex/pull/678#discussion_r1527647116

I insist on keeping the cast here. But if you know anything better (simpler) that keeps this property always true, thank I'm open to change.

```
x == decode(encode(x))
```

Reply:
Unfortunately it won't work. I already had this discussion with @emizzle here https://github.com/codex-storage/nim-codex/pull/678#discussion_r1527647116

This is the property that has to be satisfied for a successful pair of `encode` and `decode` procs:
```
let x
x == decode(encode(x))
```

This code will fail when we'll use type conversion in place of casting. It will fail for all negative values of int.

Reply:
Firstly, we need to add the appropriate missing tests then.

Reply:
I agree. I'm going to add tests for these coders.

Reply:
I would still narrow this down to only specific numeric types - `Natural | enum` and leave `int` out of the cast, because that opens this up for a lot of potential mistakes. 

This works just fine, right now:

```
proc encode*(i: Natural | enum): seq[byte] = cast[uint64](i).encode
proc decode*(T: typedesc[Natural | enum], bytes: seq[byte]): ?!T = uint64.decode(bytes).map((ui: uint64) => cast[T](ui))
```

Reply:
I've added the `testcoders.nim` that checks this property for random values of different types.

Reply:
Lets also constrain the encoding/decoding away from `int` as in my example above. Otherwise this is just way too unsafe.